lock.go

```diff
@@ -36,13 +36,18 @@ func (l *lock) TryLockWithTimeout(timeout time.Duration) bool {
 		return true
 	}
 
+	endTime := time.Now().Add(timeout)
 	// slow path
-	select {
-	case l.lockChan <- struct{}{}:
-		return true
-	case <-time.After(timeout):
-		// Failed to Acquire lock
-		return false
+	for {
+		select {
+		case l.lockChan <- struct{}{}:
+			return true
+		default:
+			// Failed to Acquire lock, check for timeout
+			if endTime.Sub(time.Now()) <= 0.0 {
+				return false
+			}
+		}
 	}
```

Comment:
I think it will burn lot of cpu cycles as it will continuously checking for timeout ( like an infinite loop before the timeout ).. i would prefer more optimised way to solve this